lib/spack/spack/externals.py

```diff
@@ -87,6 +87,7 @@ def complete_architecture(node: spack.spec.Spec) -> None:
         node.constrain(spack.spec.Spec.default_arch())
         node.architecture.target = spack.archspec.HOST_TARGET_FAMILY
 
+    node.namespace = spack.repo.PATH.repo_for_pkg(node.name).namespace
```

Comment:
FWIW, just a few lines down it goes:

```
    complete_architecture(node)
    pkg_class = spack.repo.PATH.get_pkg_class(node.name)
```

which is short for `self.repo_for_pkg(pkg_name).get_pkg_class(pkg_name)`.

Would it make sense to move it there? I think so, given that namespace has little to do with `complete_architecture`.